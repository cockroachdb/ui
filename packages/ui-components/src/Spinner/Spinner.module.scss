@import "../styles/tokens.scss";

// stroke values depend on viewBox of svg file (viewBox="0 0 100 100")
svg.spinner {
  circle {
    fill: transparent;
    stroke: $color-core-neutral-6;
    stroke-linecap: round;
    stroke-dasharray: 220;
    stroke-dashoffset: 280;
    stroke-width: 10px;
    transform-origin: 50% 50%;
    animation: spinner 3s linear infinite;
  }
}

.size-default {
  width: crl-gutters(5);
  height: crl-gutters(5);
}

.size-small {
  width: crl-gutters(3);
  height: crl-gutters(3);
}

.size-large {
  width: crl-gutters(8);
  height: crl-gutters(8);
}

@keyframes spinner {
  0% {
    stroke-dashoffset: 25;
    transform: rotate(0deg);
  }

  50% {
    stroke-dashoffset: 125;
    transform: rotate(720deg);
  }

  100% {
    stroke-dashoffset: 25;
    transform: rotate(1080deg);
  }
}
