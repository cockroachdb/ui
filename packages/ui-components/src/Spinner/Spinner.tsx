import React from "react";
import classNames from "classnames/bind";
import styles from "./Spinner.module.scss";
import objectToClassnames from "../utils/objectToClassnames";

const cx = classNames.bind(styles);

export type SpinnerSize = "default" | "large" | "small";

export interface SpinnerProps {
  size?: SpinnerSize;
  className?: string;
}

export const Spinner: React.FC<SpinnerProps> = ({
  size = "default",
  className,
}) => {
  const classNames = cx("spinner", objectToClassnames({ size }), className);
  return (
    <svg
      className={classNames}
      viewBox="0 0 100 100"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        fillRule="evenodd"
        clipRule="evenodd"
        stroke="black"
        cx="50"
        cy="50"
        r="44"
        strokeWidth="12"
        strokeDasharray="283"
        strokeDashoffset="75"
        strokeLinecap="round"
      />
    </svg>
  );
};
