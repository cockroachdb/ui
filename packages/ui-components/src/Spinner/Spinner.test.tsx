import React from "react";
import { shallow } from "enzyme";
import { Spinner, SpinnerSize } from "./Spinner";

const spinnerSizes: SpinnerSize[] = ["default", "small", "large"];

describe("Spinner", () => {
  test("renders on a screen", () => {
    const wrapper = shallow(<Spinner />);
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  spinnerSizes.forEach(size => {
    test(`renders with provided ${size} size`, () => {
      const wrapper = shallow(<Spinner size={size} />);
      expect(wrapper.hasClass(`size-${size}`)).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
