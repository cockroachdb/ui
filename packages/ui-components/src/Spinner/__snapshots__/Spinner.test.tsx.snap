// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner renders on a screen 1`] = `
<svg
  className="spinner size-default"
  fill="none"
  viewBox="0 0 100 100"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    clipRule="evenodd"
    cx="50"
    cy="50"
    fillRule="evenodd"
    r="44"
    stroke="black"
    strokeDasharray="283"
    strokeDashoffset="75"
    strokeLinecap="round"
    strokeWidth="12"
  />
</svg>
`;

exports[`Spinner renders with provided default size 1`] = `
<svg
  className="spinner size-default"
  fill="none"
  viewBox="0 0 100 100"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    clipRule="evenodd"
    cx="50"
    cy="50"
    fillRule="evenodd"
    r="44"
    stroke="black"
    strokeDasharray="283"
    strokeDashoffset="75"
    strokeLinecap="round"
    strokeWidth="12"
  />
</svg>
`;

exports[`Spinner renders with provided large size 1`] = `
<svg
  className="spinner size-large"
  fill="none"
  viewBox="0 0 100 100"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    clipRule="evenodd"
    cx="50"
    cy="50"
    fillRule="evenodd"
    r="44"
    stroke="black"
    strokeDasharray="283"
    strokeDashoffset="75"
    strokeLinecap="round"
    strokeWidth="12"
  />
</svg>
`;

exports[`Spinner renders with provided small size 1`] = `
<svg
  className="spinner size-small"
  fill="none"
  viewBox="0 0 100 100"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    clipRule="evenodd"
    cx="50"
    cy="50"
    fillRule="evenodd"
    r="44"
    stroke="black"
    strokeDasharray="283"
    strokeDashoffset="75"
    strokeLinecap="round"
    strokeWidth="12"
  />
</svg>
`;
